### DISCUSSION TITLE
[v4] Custom class extract method not working

### DISCUSSION THREAD

**Discussion by jacksleight** [tailwindlabs/tailwindcss#15176] [Help]
I'm using a custom class extract method for certain files but with v4 it throws an error if I include it in the config:

```js
export default {
    content: {
        extract: {
            php: (content) => {
                // ....
            },
        },
    },
    // ...
}
```

The error is:

> [@tailwindcss/vite:generate:build] a is not iterable

Seems it's expecting `content` to only be an array.

**Comment by adamwathan** [comment]
Hey! We don't support custom extractors in v4 right now and are really hoping we don't have to 😬 Can you share what yours was doing? Maybe it's something we can just handle?

**Reply by jacksleight** [reply]
> Maybe it's something we can just handle?

Hehe, probably only if you decided to implement something like [this](https://github.com/tailwindlabs/tailwindcss/discussions/8337). 🙂

I have an extract function and a matching PHP helper that turns strings like this:

```php
"disabled:: opacity-50 cursor-not-allowed bg-gray-100"
```

Into this:

```php
"disabled:opacity-50 disabled:cursor-not-allowed disabled:bg-gray-100"
```

It's actually part of a [Laravel package](https://github.com/jacksleight/blade-tailor#using-tailwind-variant-shorthand) I built.

The advantage of this approach is that there's no CSS bloat, since PHP converts all the classes to standard ones, it just needs a way for the TW compiler to detect them. Of course the disadvantage is it requires that support in the templating language.

At the end of the day it's not a huge issue for me if v4 wont support extract, I can always duplicate the variant prefixes like normal. I just quite like the shorter syntax (but I totally understand why you don't want to implement variant groups).

Although I wonder if there's a way that TW could support some kind of grouping syntax in the compiler, but only when used in combination with a plugin for the templating language that splits them out? Just a thought, might be a terrible idea.
